### DISCUSSION TITLE
how to import local js scripts?

### DISCUSSION THREAD

**Discussion by yanghw116** [zauberzeug/nicegui#2490] [Q&A]
### Question

hi everyone, I'm new to nicegui.

I try to import js scripts from other `.js` file,
```py
# works fine
ui.add_head_html('<script src="https://cdnjs.cloudflare.com/ajax/libs/leader-line/1.0.7/leader-line.min.js"></script>')

# not working
ui.add_head_html('<script type="text/javascript" src="./sub/leader-line.js"></script>')
```

so, if a `.js` file only exists on my disk, how to import it?

**Comment by falkoschindler** [comment]
You can serve static files using `app.add_static_files` with a given route and directory path:
https://nicegui.io/documentation/section_pages_routing#add_a_directory_of_static_files_
Then you can use the route in `ui.add_head_html`.

**Comment by rodja** [comment]
Or you load the content from file and write it directly in the head html like we do for https://nicegui.io:

https://github.com/zauberzeug/nicegui/blob/f4c57a962f8cc96085eb41860c1346da77228824/website/header.py#L13-L14

**Reply by rodja** [reply]
Although, now that I think of it: This is not async @falkoschindler. We should change this to use aiofiles.

**Reply by falkoschindler** [reply]
@rodja Even better: We should keep the files in memory. See https://github.com/zauberzeug/nicegui/commit/5e2c79ddbe96f2a38cfe463bee3dc0e0c4adb0a8.

**Reply by yanghw116** [reply]
```py
ui.add_head_html(f'''<script>{(Path(__file__).parent / 'sub' / 'leader-line.min.js').read_text()}</script>''') 
```
both are ok. this one is better?

**Reply by falkoschindler** [reply]
Loading the JavaScript file in Python and adding it to the HTML page using `add_head_html` saves you an extra HTTP request, which is good. But Python will need to open and read the file every time the page is requested (assuming you're using a `@page` function), which blocks the main thread for a short time. So it is even better to call `read_text()` once in global scope and use the result when ever `add_head_html` is called.
Note that this is just a minor optimization. For small projects without too many concurrent users all mentioned solutions should work nicely.
